### PR TITLE
dynamic estimate of required memory usage

### LINKED ISSUE #153: ggml_new_tensor_impl: not enough space in the context's memory pool (needed 717778556, available 454395136)
Hey, I know someone already posted a similar issue that has already been closed, but I ran into the same thing. On windows 10 and cloned just yesterday

### PR DESCRIPTION
**Pull request by Green-Sky**
uses observations made in #213 and replaces it.

fixes `ggml_new_tensor_impl: not enough space in the context's memory pool` and resulting Segfaults.

this is still as much of a hack as it was before, but this time it is working.

this could potentially fix a bunch of issues. ( fixes https://github.com/ggerganov/llama.cpp/issues/153 )

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -97,7 +97,9 @@ struct llama_context {
     llama_model model;
     llama_vocab vocab;
 
-    size_t mem_per_token = 0;
+	// used to estimate memory requiremnts experimentally
+    size_t mem_at_token0 = 0; // fist time
```
[the remark this answers is not in the document]
Comment (Green-Sky) [reply llama.cpp]: > Thanks. Two more typos in main.cpp: mesurements, to determain

:see_no_evil: 

> I don't fully grok that bit of the project, so I'm just nitpicking the look, sorry... Someone else will have to review as well.

yes, i intentionally requested from more then i would usually.

**Comment by Green-Sky** [comment]
hold up, need to fix perplexity.

**Comment by mqy** [comment]
@Green-Sky UB is hard to fix, I really appreciate!

I'll try this PR tomorrow. Before that, let me to make an immature suggestion:

Think about the situation that new segmentation fault  occur again, but still take time fix.
Can we add an experimental cli parameter (or env variable), to allow user configure the max memory pool . According to his/her available RAM size (e.g. at least 1 GB). This benefit users who has big RAM  and eager to evaluate.

**Comment by Green-Sky** [comment]
```
==1743825==ERROR: AddressSanitizer: allocator is out of memory trying to allocate 0x1000000000 bytes
    #0 0x7f6d6c849808 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:144
    #1 0x561e0c966487 in llama_eval_internal /home/green/workspace/github/llama.cpp/llama.cpp:636
    #2 0x561e0c9682bb in llama_eval /home/green/workspace/github/llama.cpp/llama.cpp:1460
    #3 0x561e0c879365 in main /home/green/workspace/github/llama.cpp/main.cpp:229
    #4 0x7f6d6b890082 in __libc_start_main ../csu/libc-start.c:308

==1743825==HINT: if you don't care about these errors you may set allocator_may_return_null=1
SUMMARY: AddressSanitizer: out-of-memory ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:144 in __interceptor_malloc
==1743825==ABORTING
```

so, 32GiB are not enough to run perplexity (defaults) on 7B q4_0 .
and 64GiB fail to allocate on my machine ...

**Comment by Green-Sky** [comment]
I decided to wait on https://github.com/ggerganov/llama.cpp/pull/439 and maybe https://github.com/ggerganov/llama.cpp/pull/407, so i don't break perplexity.

**Comment by Green-Sky** [comment]
for some reason @ggerganov pushed https://github.com/ggerganov/llama.cpp/commit/4870e455b3653f7d7769fa5772b2c90ffad088df :eyes:

**Comment by PerthGoat** [comment]
> for some reason @ggerganov pushed [4870e45](https://github.com/ggerganov/llama.cpp/commit/4870e455b3653f7d7769fa5772b2c90ffad088df) 👀

Unfortunately, this still doesn't fix the memory allocation issues :(

**Comment by Green-Sky** [comment]
@ggerganov promised an memory overhaul herer https://github.com/ggerganov/llama.cpp/pull/407#issuecomment-1482257457

so i am closing this pr.

**Comment by Green-Sky** [comment]
officially replaced by https://github.com/ggerganov/llama.cpp/pull/473

**Pull request closed without merging**